docs: add EnableDefaultDeny documentation (#32097)
cilium-team/cilium: Documentation/security/policy/intro.rst

The configuration of the Cilium agent and the Cilium Network Policy determines whether an endpoint accepts traffic from a source or not. The agent can be put into the following three policy enforcement modes:

default
  This is the default behavior for policy enforcement. In this mode, endpoints
  have unrestricted network access until selected by policy. Upon being selected by
  a policy, the endpoint permits only allowed traffic. This state is per-direction

>   and can be adjusted on a per-policy basis. :ref:`See below<policy_mode_default>` for
>   more details.

Reviewer: ```suggestion
  and can be adjusted on a per-policy basis. For more details, :ref:`see the dedicated section on default mode<policy_mode_default>`.
```

always
  With always mode, policy enforcement is enabled on all endpoints even if no
  rules select specific endpoints.

  If you want to configure health entity to check cluster-wide connectivity when 
  you start cilium-agent with ``enable-policy: always``, you will likely want to
  enable communications to and from the health endpoint. See :ref:`health_endpoint`.

never
  With never mode, policy enforcement is disabled on all endpoints, even if
  rules do select specific endpoints. In other words, all traffic is allowed
  from any source (on ingress) or destination (on egress).

> To :ref:`configure <k8s_configuration>` the policy enforcement mode, you can adjust the Helm value
> ``policyEnforcementMode`` or the corresponding configuration flag ``enable-policy``

Reviewer: ```suggestion
To :ref:`configure <k8s_configuration>` the policy enforcement mode, adjust the Helm value
``policyEnforcementMode`` or the corresponding configuration flag ``enable-policy``.
```

.. _policy_mode_default:

Endpoint default policy
-----------------------

By default, all egress and ingress traffic is allowed for all endpoints. When

> an endpoint is selected by a network policy, it transitions in to a default-deny

Reviewer: ```suggestion
an endpoint is selected by a network policy, it transitions to a default-deny
```

state, where only **explicitly allowed** traffic is permitted. This state is
per-direction:

* If any rule selects an :ref:`endpoint` and the rule has an ingress

>   section, the endpoint goes into default deny mode for ingress.

Reviewer: ```suggestion
  section, the endpoint goes into default-deny mode for ingress.
```

* If any rule selects an :ref:`endpoint` and the rule has an egress section, the

>   endpoint goes into default deny mode for egress.

Reviewer: ```suggestion
  endpoint goes into default-deny mode for egress.
```

This means that endpoints start without any restrictions, and the first
policy will switch the endpoint's default enforcement mode (per direction).

It is possible to create policies that do not enable the default-deny mode for selected
endpoints. The field ``EnableDefaultDeny`` configures this. Rules with ``EnableDefaultDeny``
disabled are ignored when determining the default mode.

For example, this policy causes all DNS traffic to be intercepted, but does not

> block any traffic, even if it is the first policy to apply to an endpoint.

Reviewer: What would be a use case for this rule? It could be interesting to mention one in the docs.

Author: Good point, added one.
